infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/AsyncStoragePodWatcher.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+
+/**
+ * Periodically checks ability to stop Asynchronous Storage Pod. It will periodically revise
+ * UserPreferences of all registered user and check specialized preferences. Preferences should be
+ * recorded if last workspace stopped and cleanup on start any workspace. Required preferences to
+ * initiate Asynchronous Storage Pod: {@link
+ * org.eclipse.che.api.workspace.shared.Constants.LAST_ACTIVE_INFRASTRUCTURE_NAMESPACE} : should
+ * contain last used infrastructure namespace {@link
+ * org.eclipse.che.api.workspace.shared.Constants.LAST_ACTIVE_WORKSPACE_ID} : should contain last
+ * used workspace id {@link org.eclipse.che.api.workspace.shared.Constants.LAST_ACTIVITY_TIME} :
+ * seconds then workspace stopped in the Java epoch format
+ */
+package org.eclipse.che.workspace.infrastructure.kubernetes.provision;
+
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static java.lang.Long.parseLong;
+import static org.eclipse.che.api.core.Pages.iterateLazily;
+import static org.eclipse.che.api.workspace.shared.Constants.LAST_ACTIVE_INFRASTRUCTURE_NAMESPACE;
+import static org.eclipse.che.api.workspace.shared.Constants.LAST_ACTIVITY_TIME;
+import static org.eclipse.che.workspace.infrastructure.kubernetes.provision.AsyncStorageProvisioner.ASYNC_STORAGE;
+
+import io.fabric8.kubernetes.api.model.DoneablePod;
+import io.fabric8.kubernetes.api.model.Pod;
+import io.fabric8.kubernetes.client.dsl.PodResource;
+import java.time.Clock;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import org.eclipse.che.api.core.ServerException;
+import org.eclipse.che.api.core.model.user.User;
+import org.eclipse.che.api.user.server.PreferenceManager;
+import org.eclipse.che.api.user.server.UserManager;
+import org.eclipse.che.api.workspace.server.WorkspaceRuntimes;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.spi.InternalRuntime;
+import org.eclipse.che.commons.schedule.ScheduleDelay;
+import org.eclipse.che.workspace.infrastructure.kubernetes.KubernetesClientFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
```

Comment:
Please describe the responsibility of this class in javadoc

Reply:
Added, but in wrong place https://github.com/eclipse/che/blob/1909e72e9cc052c8e88f566f98aa99b66404ecce/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/AsyncStoragePodWatcher.java#L13

Reply:
fixed